Initial draft of LIP-20 (#54)
diem/lip: lips/lip-20.md

`0xB1E55ED`. In the case of [`Libra::register_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-register_currency)
these capabilities are returned to the caller, which may store them in other structures.
The `MintCapability` and `BurnCapability` for each currency are [unique](https://github.com/libra/lip/blob/master/lips/lip-2.md#permissions); after
registration of a currency no future ones may be created.

## On-chain list of registered currencies

Whenever a currency is registered on-chain, the to-be-registered currencies currency code is added to the
[`RegisteredCurrencies` on-chain config](https://github.com/libra/lip/blob/master/lips/lip-6.md#registered-currencies).
The set of currency codes registered on-chain is always a set. Attempting to
register a currency with an already taken currency code will fail. There are no
restrictions on the format or length of currency codes, but there are
[operational restrictions](#operational_things) that should be considered from the
on-chain perspective when choosing a currency code.

---
# <a name="coin_value_representation">Representation of Value</a>
---

> Every currency that is registered on-chain holds a [`scaling_factor`](#scaling_factor)

Reviewer: This is already mentioned when scaling_factor is introduced. Remove?

### Mutative ("setters")
* [`Libra::update_lbr_exchange_rate<FromCoinType>`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-update_lbr_exchange_rate)
* [`Libra::update_minting_ability<CoinType>`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-update_minting_ability)

### Predicates
* [`Libra::is_currency<CoinType>(): bool`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-is_currency)
    - Related assertion: [`Libra::assert_is_currency<CoinType>()`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-assert_is_currency)
* [`Libra::is_SCS_currency<CoinType>(): bool`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-is_SCS_currency)
    - Related assertion: [`Libra::assert_is_SCS_currency<CoinType>()`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-assert_is_SCS_currency)
* [`Libra::is_synthetic_currency<CoinType>(): bool`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-is_synthetic_currency)

---

> # <a name="off_chain_use">Off-Chain Interactions</a>

Reviewer: Why is this titled Off-Chain?

Author: This section is really talking about the events emitted for currencies and when they're emitted, so I think this isn't a good title for this section. I'll retitle the section to clarify this.

---
lip: 20
title: Libra On-Chain Currency Management
authors: Tim Zakian, Sam Blackshear
status: Draft
type: Informational
created: 08/28/2020
---

---
# Summary
---

This LIP describes the conceptual model and implementation of currencies on the Libra
blockchain. Currencies in Libra are represented statically at the type level, each
currency with a unique Move type. A type may treated as a currency in Libra
if and only if it has previously been registered as a currency on-chain. Once a currency
is registered it remains registered forever.

> Every currency that is registered on-chain has metadata

Author: > There should be commas both before and after "i.e." and "e.g." -- you've only got commas after those. It helps if you think of spelling them out as "that is" or "for example".

Good call. I'll update :) 

> I can see why you were tempted to use a semicolon between "a SCS...fiat currency" and "or a synthetic currency..." because there are already lots of commas, but I don't think a semicolon works there. It should be a comma.

Yea...I might look in to rephrasing that. Not having a clean separation there seems not very nice.

> You should have something to separate "metadata associated with it at the time of registration" from "such as the currency-code....". A comma would work, or else you could make that a separate sentence: e.g., "That metadata includes..."

👍 

> Should "currency-code" be hyphenated? I think not.

👍 

> How about "on-chain"? In this case, it depends on the usage (and I know that this is not done consistently, but hey, I'm already writing all this about punctuation!) In a phrase like "currencies registered on chain", I don't believe a hyphen should be used. The "on chain" is a prepositional phrase described where the currencies are registered. If you refer to something like "on-chain resources", the "on-chain" is used as an adjective describing the resources. (There's surely a more precise term for this, but I forget what it is.)

Ahh! yes good point. The rule of thumb I use was given to me by a PL person: use "-" for words that need to be bound to the left... So not nearly as precise as yours 😉 

> ...and then there's the Oxford comma question (before "and other information"). I tend to use those myself and since you're an Oxford grad, I would have expected one here 

They actually just called. They're planning on revoking my degree because of this.... In all seriousness thanks for pointing it out since I am a very big fan (and user!) of the Oxford comma. I'll fix :)

On-chain, all assets with a value of `x` in a registered currency `C` are held as a "Libra Coin" resource
of type [`Libra<C>`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#resource-libra)
with its `value` field set to `x`. Every Libra Coin has a certain set of
operations that may be performed on it and other coins of the same
currency[[2]](#general_currency_operations), and every registered currency possesses a
specific set of metadata about itself[[3]](#metadata_spec) that may be
queried[[4]](#currency_ops). Additionally, none of the operations detailed in the rest of
this document may be performed on a non-registered currency (i.e., type) with the
exception of registration.

---
# <a name="motivation">Motivation</a>
---

This LIP details the representation of currencies on-chain, their
registration process (in a technical setting), the representation of assets in a specific
currency, and the operations that may be performed with registered currencies and currency
assets (or "Libra Coins").

The central requirements for a currency representation in Libra can be
distilled to a couple core requirements:

> 1. *Creation and removal*: Every on-chain currency must have a way for new

Author: Good call out. I'll update across the doc.

   coins to created and removed from the system.
2. *Privilege and control*: Creation and removal of Libra Coins of any currency

>    type must be a privileged action and tightly controlled, as every coin in

Author: Good point. I'll update to remove reserve and update with "controlled yet flexible" here.

or [`Libra::register_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-register_currency)
instantiated with the type `C` that is to be registered as a new SCS or synthetic currency respectively.

## <a name="metadata_spec">Required information for registration of a currency and metadata</a>

When a currency is registered using these functions, a number of metadata need to be
provided about the currency that is being registered. In particular, every registered
currency on-chain contains the following data, names in bold are metadata that must be
provided at the time of registration:

| Name                          | Type                                   | Mutable      | Description                                                                                                             |
| ----------------------------- | -------------------------------------- | ------------ | ----------------------------------------------------------------------------------------------------------------------- |

> | **to_lbr_exchange_rate**      | FixedPoint32                           | true         | The _rough_ exchange rate from `C` to LBR. Used only for dual attestation threshold and mempool gas price normalization   |

Author: I've added a comment to the requirements portion to mention that it needs to normalize to a selected currency/value and in Libra this is chosen to be LBR.

| preburn_value                 | u64                                    | true         | The total value of all coins in this currency waiting to be burned                                                      |
| can_mint                      | bool                                   | true         | Whether more value in this currency can be added to circulation                                                         |

All of the metadata about a specific currency `C` is held as part of the
[`CurrencyInfo<C>`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#resource-currencyinfo)
resource that is uniquely published for each registered currency type `C` under the [Libra Root](https://github.com/libra/lip/blob/master/lips/lip-2.md#roles) (see also the implementation of
[`Libra::register_SCS_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-register_scs_currency)
and
[`Libra::register_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-register_currency))
account address. We now describe each of these data and their purpose.

> ### `to_lbr_exchange_rate`

Author: I went for a table here since the `CurrencyInfo` struct contains a number of other fields related to events that I don't want to go into at this point in the document. I'm open to changing this to the full resource definition, but I do have a preference for the table though because of this. But, like I said, I'm open to changing it if you still think it should be the resource definition after this :)

fluctuations in the real-time exchange rate between the specified currency `C` and
≋LBR. Note that this is not meant to be an exact exchange rate and _should not_ be used for
determining exchange rates for value transfer between different currencies.

### `is_synthetic`

This defines whether the given currency is synthetic or an SCS. In the case of
[`Libra::register_SCS_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-register_scs_currency)
the function sets this to `false`. For
[`Libra::register_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-register_currency))
this must be provided to the function.

#### Mutability
Immutable

### <a name="scaling_factor">`scaling_factor`</a>

This defines what fraction each unit of value on-chain represents in the off-chain
representation of the currency. As an example: the on-chain representation of ≋LBR is
microlibra, therefore each unit of value on-chain represents one microlibra and the
`scaling_factor` for ≋LBR is `1,000,000` since each unit is `1/1,000,000` of an ≋LBR.

### `fractional_part`

> This defines the truncation that needs to be performed when transferring from the on-chain

Reviewer 2: "Truncation" doesn't seem like the right word for this. A client may choose to display smaller fractions, as is often done for things like stock prices. I like the phrase here "smallest denomination real-world coin". Maybe this could be phrased in terms of that, with a suggestion that values be rounded (probably not truncated) for display in those units?

Author: Good call on the truncation terminology and using "smallest denomination real-world coin" instead. I'll update and include the suggestion to round and not truncate as well.

representation to the off-chain (i.e., "real world") representation of the currency due to
the restrictions of the particular currency. For example, if an SCS `C` has a scaling factor of
`1,000,000` but the smallest denomination real-world coin for `C` is 100'th of a `C`, then
the `fractional_part` for `C` would be `100`.

#### Mutability
Immutable

### `currency_code`

> The currency code is specified at the time of registration of the currency. In addition to

Reviewer 2: Should this reference ISO 4217? It's important that currency codes be unique, but at least at first, I'm assuming that our currencies will be unofficial.

Author: Good question. I purposefully avoided being too precise here since I don't think the exact format is locked down on this yet. This also might get a bit to Sam's high-level question at the top of this PR about how Libra-implementation-specific we want this to be. My thinking for now it to leave this intentionally vague for now and circle back if we want to specify this more heavily?

Whenever a type `C` is registered as a currency, a
[`MintCapability<C>`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#resource-mintcapability)
and
[`BurnCapability<C>`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#resource-burncapability),
are created. If the currency is registered using the
[`Libra::register_SCS_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-register_scs_currency)
function, both of these capabilities are stored under the Treasury Compliance account at address
`0xB1E55ED`. In the case that the [`Libra::register_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-register_currency)
is used to register the currency both of these capabilities are returned to the caller, which may store them in other structures.
The `MintCapability` and `BurnCapability` for each currency are [unique](https://github.com/libra/lip/blob/master/lips/lip-2.md#permissions); after
registration of a currency no future mint and burn capabilities for the currency may be created.

## On-chain list of registered currencies

> Whenever a currency is registered on-chain, the to-be-registered currencies currency code is added to the

Reviewer 2: ```suggestion
Whenever a currency is registered on chain, its currency code is added to the
```

[`RegisteredCurrencies` on-chain config](https://github.com/libra/lip/blob/master/lips/lip-6.md#registered-currencies).

> The set of currency codes registered on-chain is always a set. Attempting to

Reviewer 2: "the set... is always a set": there must be a better way to say that!

Author: It can't be a false statement though! :p 

I'll update.

register a currency with an already taken currency code will fail. There are no

> restrictions on the format or length of currency codes, but there are

Reviewer 2: There are indeed some restrictions. Currently they are required to be Move identifiers, and I just proposed a PR to further require that they be composed solely of alphanumeric characters (https://github.com/libra/libra/pull/6081) with a follow-up issue to see if we can restrict use of lowercase characters (https://github.com/libra/libra/issues/6082)

Author: I'll update to include those changes!

[operational restrictions](#operational_things) that should be considered from the
on-chain perspective when choosing a currency code.

---
# <a name="coin_value_representation">Representation of Value</a>
---

All assets in a currency `Currency` with a specific value `v` are represented on-chain as a
Libra Coin value, which has the following form:

```rust
Libra<Currency> {
    value: v,
} : Libra<Currency>
```

Where the value `v` is defined in the "base units" for `Currency` as defined by
the [`scaling_factor`](#scaling_factor) for `Currency`.

It is important to note that in the system any value of type `Libra<Currency>` is considered a
valid asset in the `Currency` currency with its value given by the `value` field.  Therefore the
creation of any non-zero value of this type needs to be considered a privileged operation
since it represents creation of value on-chain. Additionally, this is an if-and-only-if
relation; a value represents a specific amount `v` in a given currency `Currency`, if, and only
if the value has the type `Libra<Currency>`.

---

> # <a name="minting_and_burning">Creation and Removal of Value</a>

Author: Agreed that adding code in will be helpful here. But, I think it's important to not go to code heavy, and I think walking a middle ground here would be beneficial. Particularly, in my thinking, since this is an informational LIP I don't want to describe "this is how it's done" but also to some extent "this is why it's done."

if and only if it has previously been registered as a currency on-chain. Once a currency
is registered it remains registered forever.

Every currency that is registered on-chain has metadata
associated with it at the time of registration. That metadata includes the currency code, value
held on-chain in the currency, and other relevant information[[1]](#registration_and_metadata). One of the most
important properties held in a currency's metadata is whether it is a Single Currency Stablecoin
(SCS), i.e., a fiat currency, or a synthetic currency that may consist of one or
multiple other currencies registered on-chain, e.g., ≋LBR.

On-chain, all assets with a value of `x` in a registered currency `C` are held as a "Libra Coin" resource
of type [`Libra<C>`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#resource-libra)
with its `value` field set to `x`. Every Libra Coin has a certain set of
operations that may be performed on it and other coins of the same
currency[[2]](#general_currency_operations), and every registered currency possesses a
specific set of metadata about itself[[3]](#metadata_spec) that may be

> queried[[4]](#currency_ops). Additionally, none of the operations detailed in the rest of

Reviewer 3: This sentence has too many negatives to grok easily :). Perhaps: "All operations must be be performed on a registered currency type"?

this document may be performed on a non-registered currency (i.e., type) with the
exception of registration.

---
# <a name="requirements">Requirements</a>
---

> This LIP details the representation of currencies on-chain, their

Reviewer 3: "representation of currencies" and "representation of assets in a specific currency" sound similar--maybe different phrasing?

registration process (in a technical setting), the representation of assets in a specific
currency, and the operations that may be performed with registered currencies and currency
assets (or "Libra Coins").

The requirements for a currency in Libra are as follows:
1. *Privileged Registration*: The registration of a Move type as a currency

>    on-chain must be a privileged operation, and only possible to accounts that have the

Reviewer 3: ```suggestion
   on-chain is restricted to accounts with the
```

or [`Libra::register_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-register_currency)
instantiated with the type `C` that is to be registered as a new SCS or synthetic currency respectively.

## <a name="metadata_spec">Required information for registration of a currency and metadata</a>

When a currency is registered using these functions, a number of metadata need to be
provided about the currency that is being registered. In particular, every registered
currency on-chain contains the following data, names in bold are metadata that must be
provided at the time of registration:

| Name                          | Type                                   | Mutable      | Description                                                                                                                                      |
| ----------------------------- | -------------------------------------- | ------------ | -----------------------------------------------------------------------------------------------------------------------                          |

> | **to_lbr_exchange_rate**      | FixedPoint32                           | true         | The _rough_ exchange rate from `C` to LBR. Used only for dual attestation threshold and mempool gas price normalization                          |

Reviewer 3: Will we still have this at launch?

Author: yep! We keep this around so that we won't need to upgrade the `CurrencyInfo` resource, and because this is used for both the dual attestation normalization and gas price normalization.

by an account with the [Treasury Compliance role](https://github.com/libra/lip/blob/master/lips/lip-2.md#roles) to correspond to
fluctuations in the real-time exchange rate between the specified currency `C` and
≋LBR. Note that this is not meant to be an exact exchange rate and _should not_ be used for
determining exchange rates for value transfer between different currencies.

### `is_synthetic`

This defines whether the given currency is synthetic or an SCS. In the case of
[`Libra::register_SCS_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-register_scs_currency)
the function sets this to `false`. For
[`Libra::register_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-register_currency)
this must be provided to the function.

#### Mutability
Immutable

### <a name="scaling_factor">`scaling_factor`</a>

This defines what fraction each unit of value on-chain represents in the off-chain
representation of the currency. As an example: the on-chain representation of ≋LBR is
microlibra, therefore each unit of value on-chain represents one microlibra and the

> `scaling_factor` for ≋LBR is `1,000,000` since each unit is `1/1,000,000` of an ≋LBR.

Reviewer 3: This example is extremely helpful! But would be good to use the same dummy `C` type for consistency with the `fractional_part` example.

        Libra::assert_is_currency<C>()
    ```
2. Return if the type `C` represents a registered SCS currency on-chain.
    ```rust
        Libra::is_SCS_currency<C>(): bool
    ```
    Related assertion:
    ```rust
        Libra::assert_is_SCS_currency<C>()
    ```
3. Return if the type `C` represents a registered synthetic currency on-chain.
    ```rust
        Libra::is_synthetic_currency<C>(): bool
    ```

#### Reference Implementations

* [`Libra::is_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-is_currency)
    - Related assertion [`Libra::assert_is_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-assert_is_currency)
* [`Libra::is_SCS_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-is_SCS_currency)
    - Related assertion [`Libra::assert_is_SCS_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-assert_is_SCS_currency)
* [`Libra::is_synthetic_currency`](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#function-is_synthetic_currency)

---

> # <a name="currency_events">Currency-Related Events</a>

Reviewer 3: If it's not too much work, could be useful to discuss reconciliation of events here. E.g., 

sum(values in all mint events for C) - sum(values in all burn events for C) = total value for C

| cancel_burn_events            | EventHandle\<[CancelBurnEvent](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#struct-cancelburnevent)\>                 | Event emitted whenever SCSs in a preburn area are moved back to circulation                                 |
| exchange_rate_update_events   | EventHandle\<[ExchangeRateUpdateEvent](https://github.com/libra/libra/blob/master/language/stdlib/modules/doc/Libra.md#struct-exchangerateupdateevent)\> | Event emitted whenever the `to_lbr_exchange_rate` metadata field is updated (for all registered currencies) |

---
# <a name="operational_things">Operational Restrictions</a>
---

When deciding the on-chain currency code for a currency,
the module, as well as the type name chosen for the currency must match the on-chain currency
code chosen for the currency. e.g., if you want to use currency code `"ABC"` for a
currency, then the fully qualified Move type that is used for that currency on-chain
_must_ be `0x1::ABC::ABC`.

Because of this, and in addition:
* An on-chain currency code must be a valid Move identifier; and

> * It can consist solely of alphanumeric (TODO: [uppercase](https://github.com/libra/libra/issues/6082)) characters.

Reviewer 3: Are we planning to resolve this TODO when we do the coin rename?

Author: yep! Although I think I'll remove the TODO since it isn't really adding anything IMO
